lxd/storage/utils.go

```diff
@@ -0,0 +1,406 @@
+	}
+
+	return changedConfig, userOnly
+}
+
+// Default permissions for folders in ${LXD_DIR}
+const storagePoolsDirMode os.FileMode = 0711
+const containersDirMode os.FileMode = 0711
+const customDirMode os.FileMode = 0711
+const imagesDirMode os.FileMode = 0700
+const snapshotsDirMode os.FileMode = 0700
+
+// Detect whether LXD already uses the given storage pool.
+func lxdUsesPool(dbObj *db.Cluster, onDiskPoolName string, driver string, onDiskProperty string) (bool, string, error) {
+	pools, err := dbObj.StoragePools()
+	if err != nil && err != db.ErrNoSuchObject {
+		return false, "", err
+	}
+
+	for _, pool := range pools {
+		_, pl, err := dbObj.StoragePoolGet(pool)
+		if err != nil {
+			continue
+		}
+
+		if pl.Driver != driver {
+			continue
+		}
+
+		if pl.Config[onDiskProperty] == onDiskPoolName {
+			return true, pl.Name, nil
+		}
+	}
+
+	return false, "", nil
+}
+
+func makeFSType(path string, fsType string, options *mkfsOptions) (string, error) {
+	var err error
+	var msg string
+
+	fsOptions := options
+	if fsOptions == nil {
+		fsOptions = &mkfsOptions{}
+	}
+
+	cmd := []string{fmt.Sprintf("mkfs.%s", fsType), path}
+	if fsOptions.label != "" {
+		cmd = append(cmd, "-L", fsOptions.label)
+	}
+
+	if fsType == "ext4" {
+		cmd = append(cmd, "-E", "nodiscard,lazy_itable_init=0,lazy_journal_init=0")
+	}
+
+	msg, err = shared.TryRunCommand(cmd[0], cmd[1:]...)
+	if err != nil {
+		return msg, err
+	}
+
+	return "", nil
+}
+
+func fsGenerateNewUUID(fstype string, lvpath string) (string, error) {
+	switch fstype {
+	case "btrfs":
+		return btrfsGenerateNewUUID(lvpath)
+	case "xfs":
+		return xfsGenerateNewUUID(lvpath)
+	}
+
+	return "", nil
+}
+
+func xfsGenerateNewUUID(lvpath string) (string, error) {
+	msg, err := shared.RunCommand(
+		"xfs_admin",
+		"-U", "generate",
+		lvpath)
+	if err != nil {
+		return msg, err
+	}
+
+	return "", nil
+}
+
+func btrfsGenerateNewUUID(lvpath string) (string, error) {
+	msg, err := shared.RunCommand(
+		"btrfstune",
+		"-f",
+		"-u",
+		lvpath)
+	if err != nil {
+		return msg, err
+	}
+
+	return "", nil
+}
+
+func growFileSystem(fsType string, devPath string, mntpoint string) error {
+	var msg string
+	var err error
+	switch fsType {
+	case "": // if not specified, default to ext4
+		fallthrough
+	case "ext4":
+		msg, err = shared.TryRunCommand("resize2fs", devPath)
+	case "xfs":
+		msg, err = shared.TryRunCommand("xfs_growfs", devPath)
+	case "btrfs":
+		msg, err = shared.TryRunCommand("btrfs", "filesystem", "resize", "max", mntpoint)
+	default:
+		return fmt.Errorf(`Growing not supported for filesystem type "%s"`, fsType)
+	}
+
+	if err != nil {
+		errorMsg := fmt.Sprintf(`Could not extend underlying %s filesystem for "%s": %s`, fsType, devPath, msg)
+		logger.Errorf(errorMsg)
+		return fmt.Errorf(errorMsg)
+	}
+
+	logger.Debugf(`extended underlying %s filesystem for "%s"`, fsType, devPath)
+	return nil
+}
+
+func shrinkFileSystem(fsType string, devPath string, mntpoint string, byteSize int64) error {
```

Comment:
Defined but not used in commit

Reply:
This is used for other backends which originally were part of this PR.